rasa_core/featurizers.py

```diff
@@ -163,18 +136,425 @@ def encode(self, active_features, input_feature_map):
+
+    def training_states_and_actions(
+            self,
+            trackers,  # type: List[DialogueStateTracker]
+            domain  # type: Domain
+    ):
+        # type: (...) -> Tuple[List[List[Dict]], List[List[Dict]], Dict]
+
+        trackers_as_states = []
+        trackers_as_actions = []
+        events_metadata = defaultdict(set)
+
+        for tracker in trackers:
+            states = domain.states_for_tracker_history(tracker)
+            idx = 0
+            for event in tracker._applied_events():
+                if isinstance(event, ActionExecuted):
+                    if not event.unpredictable:
+                        # only actions which can be predicted at a stories start
+                        sliced_states = domain.slice_feature_history(
+                            states[:idx + 1], self.max_history)
+                        trackers_as_states.append(sliced_states)
+                        trackers_as_actions.append(event.action_name)
+                    idx += 1
+                else:
+                    action_name = tracker.latest_action_name
+                    events_metadata[action_name].add(event)
+
+        if self.remove_duplicates:
+            logger.debug("Got {} action examples."
+                         "".format(len(trackers_as_actions)))
+            (trackers_as_states,
+             trackers_as_actions) = self._remove_duplicate_states(
+                                        trackers_as_states,
+                                        trackers_as_actions)
+            logger.debug("Deduplicated to {} unique action examples."
+                         "".format(len(trackers_as_actions)))
+
+        # TODO do we need that?
+        metadata = {"events": events_metadata,
+                    "trackers": trackers}
+        return trackers_as_states, trackers_as_actions, metadata
+
+    def prediction_states(self,
+                          trackers,  # type: List[DialogueStateTracker]
+                          domain  # type: Domain
+                          ):
+        # type: (...) -> List[List[Dict[Text, float]]]
+
+        trackers_as_states = [domain.states_for_tracker_history(tracker)
+                              for tracker in trackers]
+        trackers_as_states = [domain.slice_feature_history(states,
+                                                           self.max_history)
+                              for states in trackers_as_states]
+
+        return trackers_as_states
+
+    @staticmethod
+    def _remove_duplicate_states(
+            trackers_as_states,  # type: List[List[Dict[Text, float]]]
+            trackers_as_actions,  # type: List[List[Dict[Text, float]]]
+    ):
+        # type: (...) -> Tuple[List[List[Dict]], List[List[Dict]]]
+        """Removes states that create equal featurizations.
+
+        From multiple states that create equal featurizations
+        we only need to keep one."""
+
+        hashed_featurizations = []
```

Comment:
should be a set

Reply:
in generator as well then:
https://github.com/RasaHQ/rasa_core/blob/031df4c36b48fc26e9bc2765c1394ad38ff2c50f/rasa_core/training/generator.py#L58